pkg/client/core/service_application.go

```diff
@@ -0,0 +1,154 @@
+package core
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/oslokommune/okctl/pkg/spinner"
+
+	kaex "github.com/oslokommune/kaex/pkg/api"
+	"github.com/oslokommune/okctl/pkg/api"
+	"github.com/oslokommune/okctl/pkg/client"
+	clientFilesystem "github.com/oslokommune/okctl/pkg/client/core/store/filesystem"
+	"github.com/oslokommune/okctl/pkg/client/store"
+	"github.com/oslokommune/okctl/pkg/scaffold"
+)
+
+type applicationService struct {
+	spin   spinner.Spinner
+	paths  clientFilesystem.Paths
+	cert   client.CertificateService
+	store  client.ApplicationStore
+	report client.ApplicationReport
+}
+
+func createCertificateFn(ctx context.Context, certService client.CertificateService, id *api.ID, hostedZoneID string) func(domain string) (string, error) {
+	return func(fqdn string) (string, error) {
+		cert, certFnErr := certService.CreateCertificate(ctx, api.CreateCertificateOpts{
+			ID:           *id,
+			FQDN:         fqdn,
+			Domain:       fqdn,
+			HostedZoneID: hostedZoneID,
+		})
+		if certFnErr != nil {
+			return "", certFnErr
+		}
+
+		return cert.CertificateARN, nil
+	}
+}
+
+func writeSuccessMessage(writer io.Writer, applicationName string, argoCDResourcePath string) {
+	fmt.Fprintf(writer, "Successfully scaffolded %s\n", applicationName)
+	fmt.Fprintln(writer, "To deploy your application:")
+	fmt.Fprintln(writer, "\t1. Commit and push the changes done by okctl")
+	fmt.Fprintf(writer, "\t2. Run kubectl apply -f %s\n", argoCDResourcePath)
+	fmt.Fprintf(writer, "If using an ingress, it can take up to five minutes for the routing to configure")
+}
+
+// ScaffoldApplication turns a file path into Kubernetes resources
+func (s *applicationService) ScaffoldApplication(ctx context.Context, opts *client.ScaffoldApplicationOpts) error {
```

Comment:
Let's say this crashes in one of the steps, and the user attempts to scaffold the application again - is this function idempotent (and thus continuable, doesn't create a new set of certificats or other resources that needs to be cleaned up)?

Reply:
Yeah pretty much. I know certificates are only created once, as long as the URL doesn't change.

I guess if writing the argocd resource should fail, the k8s resources will still have been written. I'm not about to implement transactions if thats what you're getting at :P (unless there is a function for that somewhere)

Reply:
No we don't want the complexity of that (now and here at least)! As long as it's okay to run this function 10 times, it's good.